### ISSUE TITLE
Corrupted XML file when working with a particular ideogram.

### ISSUE THREAD

**Issue by flacotelli** [abap2xlsx/abap2xlsx#298]
Hi,

I want to report what i suppose to be a bug. If i try to write in a cell a string containing a particular well defined ideogram (http://graphemica.com/%F0%A3%98%BA refer to this site for further infos about the ideogram encoding). Excel can't open the file, while open office calc opens it (but the particolar ideogram is not displayed). With a further investigation (i.e. renaming the xlsx to zip and looking the xml files) i noticed that the error is probably caused by a corrupted sharedstrings.xml file. In particulary, opening this file with a text editor, i noticed that there's an alphanumeric string instead of the actual ideogram.
![screensequenzachar](https://f.cloud.github.com/assets/6791522/2267895/c7510e58-9ec0-11e3-9ebd-b3c2bbd749cb.png)

The class seems to handle fine all the other ideograms. Any idea?Thank you

**Comment by ivanfemia** [comment]
Hi,

is the same working in Excel from scratch?
Could you please provide us a working excel file and a bad one?

Thanks
Ivan

**Comment by flacotelli** [comment]
Hello Ivan,

thanks for your response. I attach two excels files as you requested. Both
are generated by the same abap report (basically a little edited of one of
the zdemo reports).

If you have any idea or solution please let me know.

Thanks, Roberto
Il 26/feb/2014 20:30 "Ivan" notifications@github.com ha scritto:

**Comment by gregorwolf** [comment]
Dear Roberto,

I can't see the attachments. But what @ivanfemia ment is if you could provide a XLSX file that was not created by the abap2xlsx but directly in Excel. That would show us where we have to look.

Best regards
Gregor

**Comment by flacotelli** [comment]
Hi Gregor,
do you mean an xlsx file created in excel containing that particular
ideogram '
Best regards, roberto

**Comment by gregorwolf** [comment]
Yes.

**Comment by flacotelli** [comment]
Hi Gregor, I'm sorry but at the moment i don't have microsoft office, so i
can't create the xlsx file you requested. Btw if you wanna try, the damn
ideogram is the following: http://graphemica.com/%F0%A3%98%BA
If that's not enough in a couple of days at worst i will be able to send
you the file you requested.

Please kindly keep me updated as I'm very curious about this very strange
problem :-)

Big Thanks
Roberto

**Comment by flacotelli** [comment]
Hello Gregor,
if you give me your email address i'll send you the excel file you and ivan
requested.

thankx, roberto

**Comment by flacotelli** [comment]
update: https://www.dropbox.com/l/o7NfakgtJcYk1UMsGOX9Ve you can find here a working xlsx file generated from excel (with the same ideogram) and a not working file generated by abap2xlsx.

thanks

**Comment by ivanfemia** [comment]
I'll have a look in the next days

**Comment by ivanfemia** [comment]
Hi Roberto,

I checked the ideogram and I suspect that the issue is with the encoding and the ABAP native XML statements.
Could you please try to change it in UTF-16 in the writer class and let me know?

I created this simple example abap report and using UTF-8 is not working using UTF-16 it works.
https://dl.dropboxusercontent.com/u/2262983/ABAP.abap

Sound strange to me that Excel is able to use UTF-8 xml. Some further investigation is needed.

**Comment by flacotelli** [comment]
Hi Ivan, I already suspected that the problem was in the standard ixml
class, i created a sample report using only standard ixml classes and got
to your same result (i confirm that setting the encoding to uft-16 solves
the issue). My question now is: is there something that can be done from
abap2xlsx side to make the xml created in utf-16 or should i "modify" the
standard ixml class itself ?
Thanks, Roberto

**Comment by ivanfemia** [comment]
encoding is defined in the ZCL_EXCEL_WRITER_2007 class, check each method.
You can try to change it to UTF-16 and verify.

What I can't figure out is that standard Excel uses UTF-8, I had no time now to investigate, but i hope that this quick fix will solve your issue...

**Comment by flacotelli** [comment]
Hi Ivan, i confirm you that changing the encoding to uft-16 in the writer
class solves the issue. As we suspected the problem was in the standard
ixml class that abap2xlsx uses to write xml files. Instead of changing all
the methods of the z_writer class maybe for the next releases it could be
useful to set the xml encoding as a global attribute.

Big thanks for the support !

**Comment by gregorwolf** [comment]
Seems that the same issue as found with another special character. Find my sample code at http://scn.sap.com/thread/3801086 showing problems with UTF-8 using ixml.

**Comment by mlaggner** [comment]
@gregorwolf we've hit the same issue today and played around with your example from above.
After visiting the notes you have put into your article, we had success filtering out those illegal characters. According to the note [1750204 - iXML: Treatment of special and invalid characters](http://service.sap.com/sap/support/notes/1750204) there is a new method in the interface `IF_IXML_OSTREAM` which enables skipping of _invalid_ characters: `SKIP_NON_XML_CHARACTERS`

we've modified your test program to make it work with the hints of this note:

```
  DATA(lo_streamfactory) = lo_ixml->create_stream_factory( ).
  DATA(lo_ostream) = lo_streamfactory->create_ostream_xstring( string = lv_content ).

  " tell the iXML engine to skip those illegal characters
  lo_ostream->skip_non_xml_characters( ).

  DATA(lo_renderer) = lo_ixml->create_renderer(
                        ostream  = lo_ostream
                        document = lo_document
                      ).
  lo_renderer->render( ).
```

Having a look at the class `ZCL_EXCEL_WRITER_2007` we have found the same line of code in the method `render_xml_document` https://github.com/sapmentors/abap2xlsx/blob/master/src/zcl_excel_writer_2007.clas.abap#L7820

I think that we could patch this method to fix that issue?
